crates/sui-graphql-rpc/src/types/balance.rs

```diff
@@ -80,27 +79,35 @@ impl Balance {
         db: &Db,
         page: Page<Cursor>,
         address: SuiAddress,
+        checkpoint_viewed_at: Option<u64>,
     ) -> Result<Connection<String, Balance>, Error> {
-        let (prev, next, results) = db
-            .execute(move |conn| {
-                page.paginate_query::<StoredBalance, _, _, _>(conn, move || {
-                    use objects::dsl;
-                    dsl::objects
-                        .select((
-                            sql::<Nullable<Text>>("CAST(SUM(coin_balance) AS TEXT)"),
-                            sql::<Nullable<SqlBigInt>>("COUNT(*)"),
-                            dsl::coin_type.assume_not_null(),
-                        ))
-                        .filter(dsl::owner_id.eq(address.into_vec()))
-                        .filter(dsl::owner_type.eq(OwnerType::Address as i16))
-                        .filter(dsl::coin_type.is_not_null())
-                        .group_by(dsl::coin_type)
-                        .into_boxed()
-                })
+        let response = db
+            .execute_repeatable(move |conn| {
+                let (lhs, mut rhs) = Checkpoint::available_range(conn)?;
+
+                if let Some(checkpoint_viewed_at) = checkpoint_viewed_at {
+                    if checkpoint_viewed_at < lhs || rhs < checkpoint_viewed_at {
+                        return Ok(None);
+                    }
+                    rhs = checkpoint_viewed_at;
+                }
+
+                let result = page.paginate_raw_query::<StoredBalance>(conn, move || {
+                    Balance::base_query(address, None, lhs as i64, rhs as i64)
+                })?;
+
+                Ok::<_, diesel::result::Error>(Some(result))
```

Comment:
I'm surprised that you need to annotate the error here -- what happens when you leave that off?

Reply:
Yeah, I have no idea what's going on here either -- there's even prior calls to `Ok` in the same function?  Only suggestion I have is to try a slightly different pattern:

```rust
page.paginate_raw_query::<StoredBalance>(
    conn,
    balance_query(address, None, lhs as i64, rhs as i64),
)
.map(Some);
```

Reply:
Ah yeah that did the trick. It's because the first one returns Result<None>, while the second one is just T, so dropping the `?` and adding the `.map(Some)` conforms to `Result<Option<T>>`